### PR TITLE
feat: add dapr addon

### PR DESCRIPTION
**Pull request by chwetion**
add dapr addon fixed https://github.com/oam-dev/kubevela/issues/609

since I don't use dapr much, there are not too many user story now, please create issue in the community, I will updated this addon continually

### CODE REVIEW

File: experimental/addons/dapr/metadata.yaml [hunk 1]
```diff
@@ -0,0 +1,13 @@
+name: dapr
+version: 1.0.0
+description: Distributed application sidecar provider standard interface
+icon: https://github.com/dapr/dapr/blob/master/img/dapr_logo.svg
+url: https://dapr.io/
+
+tags:
+- data-mesh
+- standard-interface
+
```
Comment (wangyikewxgm) [inline experimental/addons/dapr/metadata.yaml:10]: Should depend on fluxcd ？

Comment (chwetion) [reply experimental/addons/dapr/metadata.yaml:10]: yes, has fixed

File: experimental/addons/dapr/definitions/pub-trait-def.yaml [hunk 2]
```diff
@@ -0,0 +1,18 @@
+apiVersion: core.oam.dev/v1beta1
+kind: TraitDefinition
+metadata:
+  name: pub
```
Comment (barnettZQG) [inline experimental/addons/dapr/definitions/pub-trait-def.yaml:4]: is it distinct using dapr-pub？

Comment (wonderflow) [reply experimental/addons/dapr/definitions/pub-trait-def.yaml:4]: yes, better to be dapr-publish

File: experimental/addons/dapr/definitions/sub-trait-def.yaml [hunk 3]
```diff
@@ -0,0 +1,21 @@
+kind: TraitDefinition
+metadata:
+  name: sub
```
Comment (wonderflow) [inline experimental/addons/dapr/definitions/sub-trait-def.yaml:3]:
```suggestion
  name: dapr-subscribe
```

File: experimental/addons/dapr/readme.md [hunk 4]
```diff
@@ -0,0 +1,3 @@
+# dapr
+
+This addon provides dapr support for pubsub.
```
Comment (wonderflow) [inline experimental/addons/dapr/readme.md:3]:
```suggestion
This addon provides dapr support for pub/sub.
```

Can you proviode more introduction about this addon? let's add some examples about how to use and add some links.

**Comment by cezhang** [comment]
I think the pub-sub trait not quiet finished yet . For dapr, pub-sub should have three components 
- publish 
- message broker
- subscription

I guess the publish code will be included in the "up-coming" Dapr trait controller, and the message broke should be defined by dapr component. Since subscription can be defined  "Declaratively" or "Programmatically"(https://docs.dapr.io/developing-applications/building-blocks/pubsub/howto-publish-subscribe/). So overall, better to provide ways for user to defines dapr component and dapr subscription.

BTW, I like to help with this feature and the planned "Observability" feature, any group I can joined to have more detail?

**Comment by wonderflow** [comment]
> I think the pub-sub trait not quiet finished yet . For dapr, pub-sub should have three components
> 
> * publish
> * message broker
> * subscription
> 
> I guess the publish code will be included in the "up-coming" Dapr trait controller, and the message broke should be defined by dapr component. Since subscription can be defined "Declaratively" or "Programmatically"(https://docs.dapr.io/developing-applications/building-blocks/pubsub/howto-publish-subscribe/). So overall, better to provide ways for user to defines dapr component and dapr subscription.
> 
> BTW, I like to help with this feature and the planned "Observability" feature, any group I can joined to have more detail?

Would you like to join our DingTalk? You can add my DingTalk with ID `wonderflow`

**Comment by chwetion** [comment]
@cezhang does message broker refer to component resources? if so I have considered two scenarios:
1. The component is also in kubevela, so that component resource can be a trait for component workload
2. The component is outside the k8s cluster, i don't have a good idea to use the trait to create the component resource

**Comment by cezhang** [comment]
> @cezhang does message broker refer to component resources? if so I have considered two scenarios:
> 
> 1. The component is also in kubevela, so that component resource can be a trait for component workload
> 2. The component is outside the k8s cluster, i don't have a good idea to use the trait to create the component resource

No matter broker inside or outside kubevela, the pub-sub feature from dapr must define a "dapr component" for message broker, like this one 
```
apiVersion: dapr.io/v1alpha1
kind: Component
metadata:
  name: rabbitmq
spec:
  type: pubsub.rabbitmq
  metadata:
  - name: host
    value: amqp://localhost:5672
```

I was wondering when I enable this trait for my workload, should I manually setup this thing?  I guess this setup should be embraced into this trait  and expose argument for users to configurate theirs message broker.

**Comment by chwetion** [comment]
@cezhang I think you are right, I will add it, BTW, do users need to fill in the address twice on pub and sub

**Comment by cezhang** [comment]
@chwetion 

I suggest that  the pub-trait should include "message broker" dapr component,  
the sub-trait includes  declarative subscription like the following example, so the end user no need to change code to subscribe a topic. But anyway no harm to provide one more option for end users.
 For reference, https://docs.dapr.io/developing-applications/building-blocks/pubsub/howto-publish-subscribe/ "Step 2: Subscribe to topics"
```
apiVersion: dapr.io/v1alpha1
kind: Subscription
metadata:
  name: myevent-subscription
spec:
  topic: newOrder
  route: /orders
  pubsubname: rabbitmq
scopes:
- app
```
----------
BTW, I was trying to model the dapr building blocks in cuelang, maybe a little helpful for this feature
```
"dapr-component": {
	annotations: {}
	attributes: workload: {
		type: "autodetects.core.oam.dev"
	}
	description: "Component describes an Dapr component type."
	labels: {}
	type: "component"
}

template: {
	output: {
		{
				apiVersion: "dapr.io/v1alpha1"
				kind:       "Component"
		} & parameter.component
	}
	parameter: {
		// Component describes an Dapr component type.
		component: {
			// +optional
			metadata?: #ObjectMeta

			// +optional
			spec?: #ComponentSpec

			// +optional
			auth?: #Auth

			// +optional
			scopes?: [...string]
		}

		// ComponentSpec is the spec for a component.
		#ComponentSpec: {
			type:    string
			version: string

			// +optional
			ignoreErrors?: bool
			metadata: [...#MetadataItem]

			// +optional
			initTimeout?: string
		}

		// MetadataItem is a name/value pair for a metadata.
		#MetadataItem: {
			name: string

			// +optional
			value?: #DynamicValue

			// +optional
			secretKeyRef?: #SecretKeyRef
		}

		// SecretKeyRef is a reference to a secret holding the value for the metadata item. Name is the secret name, and key is the field in the secret.
		#SecretKeyRef: {
			name: string
			key:  string
		}

		// Auth represents authentication details for the component.
		#Auth: {
			secretStore: string
		}

		// DynamicValue is a dynamic value struct for the component.metadata pair value.
		#DynamicValue: _

		#ObjectMeta: {
				name?:                       string
				namespace?:                  string
				deletionGracePeriodSeconds?: null | int64
				labels?: {
						[string]: string
				}
				annotations?: {
						[string]: string
				}
				finalizers?: [...string]
				clusterName?: string
		}
	}
}

```
```
"dapr-subscription": {
	annotations: {}
	attributes: workload: {
		type: "autodetects.core.oam.dev"
	}
	description: "Subscription describes an pub/sub event subscription."
	labels: {}
	type: "component"
}

template: {
	output: {
		{
			apiVersion: "dapr.io/v1alpha1"
			kind:       "Subscription"
		} & parameter.subscription
	}
	parameter: {
		// Subscription describes an pub/sub event subscription.
		subscription: {
			// +optional
			metadata?: #ObjectMeta
			spec?:     #SubscriptionSpec

			// +optional
			scopes?: [...string]
		}

		// SubscriptionSpec is the spec for an event subscription.
		#SubscriptionSpec: {
			topic:      string
			pubsubname: string

			// +optional
			metadata?: {[string]: string}
			route: string
		}

		#ObjectMeta: {
			name?:                       string
			namespace?:                  string
			deletionGracePeriodSeconds?: null | int64
			labels?: {
				[string]: string
			}
			annotations?: {
				[string]: string
			}
			finalizers?: [...string]
			clusterName?: string
		}
	}
}

```

**Comment by wonderflow** [comment]
Hi, @chwetion can you provide some example apps for using this addon.

File: experimental/addons/dapr/definitions/dapr-publish-message-broker-trait-def.yaml [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+  - deployments.apps
+  podDisruptive: true
+  schematic:
+    cue:
+      template: |
+        parameter: {
+          name: string
+          type: string
+          version: string
+          metadata: [string]: string
```
Comment (cezhang) [inline experimental/addons/dapr/definitions/dapr-publish-message-broker-trait-def.yaml:17]: metadata may not support case like the following:
```
apiVersion: dapr.io/v1alpha1
kind: Component
metadata:
  name: pubsub
  namespace: default
spec:
  metadata:
  - name: redisHost
    value: redis-master:6379
  - name: redisPassword
    secretKeyRef:
      key: redis-password
      name: redis
  type: pubsub.redis
  version: v1
```
metadata item is not just string key-value pair , should be like this 
```
#MetadataItem: {
	name: string

	// +optional
	value?: #DynamicValue

	// +optional
	secretKeyRef?: #SecretKeyRef
}

// SecretKeyRef is a reference to a secret holding the value for the metadata item. Name is the secret name, and key is the field in the secret.
#SecretKeyRef: {
	name: string
	key:  string
}
```

**Review by cezhang — COMMENTED** [review]
except metadata part, LGTM

**Review by wangyikewxgm — COMMENTED** [review]
Since the experimental addon wouldn't enabled by default, please test it like `argocd`:https://github.com/kubevela/catalog/blob/c12b05ac6b3174dc4e55be294b2342994e369c3b/test/e2e-test/hack/addon-vela-test.sh#L31

**Review by cezhang — COMMENTED** [review]
LGTM

**Review by wonderflow — APPROVED** [review]
thanks @cezhang @chwetion

**Pull request merged**